engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java

```diff
@@ -0,0 +1,189 @@
+package org.enso.compiler.core;
+
+import org.enso.compiler.core.ir.DiagnosticStorage;
+import org.enso.compiler.core.ir.Expression;
+import org.enso.compiler.core.ir.IdentifiedLocation;
+import org.enso.compiler.core.ir.MetadataStorage;
+import org.enso.syntax.text.Debug;
+import com.oracle.truffle.api.source.Source;
+import scala.Option;
+import scala.collection.immutable.List;
+import scala.collection.immutable.$colon$colon$;
+
+import java.io.Serializable;
+import java.util.UUID;
+import java.util.function.Function;
+
+/** [[IR]] is a temporary and fairly unsophisticated internal representation
+ * format for Enso programs.
+ *
+ * It is a purely tree-based representation to support basic desugaring and
+ * analysis passes that do not rely on the ability to create cycles in the IR
+ * itself. Its existence is the natural evolution of the older AstExpression
+ * format used during the initial development of the interpreter.
+ *
+ * Please note that all extensions of [[IR]] must reimplement `copy` to keep
+ * the id intact when copying nodes. The copy implementation should provide a
+ * way to set the id for the copy, but should default to being copied. Care
+ * must be taken to not end up with two nodes with the same ID. When using
+ * `copy` to duplicate nodes, please ensure that a new ID is provided.
+ *
+ * See also: Note [IR Equality and hashing]
+ */
+public interface IR extends Serializable {
```

Comment:
Each serializable object shall have its `serialVersionUID`. Otherwise any changes (adding method) would result in an incompatible format and inability to read previous data.